controller/src/main/java/io/pravega/controller/store/stream/InMemoryStreamMetadataStore.java

```diff
@@ -50,18 +50,21 @@
     @Override
     @Synchronized
     Stream newStream(String scope, String name) {
-        Stream stream = streams.get(scopedStreamName(scope, name));
-        if (stream != null) {
-            return stream;
+        if (streams.containsKey(scopedStreamName(scope, name))) {
+            return streams.get(scopedStreamName(scope, name));
         } else {
-            return new InMemoryStream.NonExistentStream(scope, name);
+            return new InMemoryStream(scope, name);
         }
     }
 
     @Override
     @Synchronized
     Scope newScope(final String scopeName) {
-        return scopes.get(scopeName);
+        if (scopes.containsKey(scopeName)) {
```

Comment:
You are wrong. The code as written could result in incorrect behavior. If you don't synchronize you don't merely see an out of date object you may see an inconsistent one, or an uninitialized one. Hence you could see an npe or go into an infinite loop. In some cases it could even cause a segfault. As an example please actually read http://mailinator.blogspot.com/2009/06/beautiful-race-condition.html?m=1
(I have sent you this before)